src/parakeet/Pro/Driver.cpp

```diff
@@ -260,76 +217,56 @@ namespace Pro
 
     bool Driver::isDataSmoothingEnabled()
     {
-        throwExceptionIfNotConnected();
+        assertIsConnected();
 
         return sensorConfiguration.dataSmoothing;
     }
 
     bool Driver::isDragPointRemovalEnabled()
     {
-        throwExceptionIfNotConnected();
+        assertIsConnected();
 
         return sensorConfiguration.dragPointRemoval;
     }
 
     bool Driver::isIntensityDataEnabled()
     {
-        throwExceptionIfNotConnected();
+        assertIsConnected();
 
         return sensorConfiguration.intensity;
     }
 
     Driver::ScanningFrequency Driver::getScanningFrequency_Hz()
     {
-        throwExceptionIfNotConnected();
+        assertIsConnected();
 
         return sensorConfiguration.scanningFrequency_Hz;
     }
 
-    double Driver::getScanRate_Hz()
+    void Driver::serialUpdateThreadFunction()
     {
-        std::chrono::milliseconds currentTime = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
-
-        auto ms = (currentTime - interfaceThreadStartTime).count();
-
-        if(ms == 0)
+        // modifying the baud rate of serial port can cause the connected state to be off for a brief moment
+        if (!serialPort.isConnected())
         {
-            return 0;
+            return;
         }
 
-        return interfaceThreadFrameCount / (ms / 1000.0);
-    }
-
-    void Driver::serialInterfaceThreadFunction()
-    {
-        unsigned char* line = new unsigned char[MESSAGE_DATA_BUFFER_SIZE];
-        int len = 0;
+        int charsRead = serialPort.read(serialPortDataBuffer, serialPortDataBufferLength, SERIAL_MESSAGE_DATA_BUFFER_SIZE);
```

Comment:
Seems like we could be using the newly added BufferData struct in a couple places in here. Let's add a YT issue to review this code and see if it makes sense.

Reply:
YT issue created.